lib/SILOptimizer/Mandatory/SendNonSendable.cpp

```diff
@@ -633,6 +666,352 @@ class BlockPartitionState {
+    switch (localReason.kind) {
+    case LocalConsumedReasonKind::LocalConsumeInst:
+      // there is a local consume in the pred block
+      consumedReason.addConsumeOp(localReason.localInst.value(), distance);
+      break;
+    case LocalConsumedReasonKind::LocalNonConsumeInst:
+      // ignore this case, that instruction will initiate its own search
+      // for a consume op
+      break;
+    case LocalConsumedReasonKind::NonLocal:
+      consumedReason.addOtherReasonAtDistance(
+          // recursive call
+          findConsumedAtEntryReason(SILBlock, consumedVal), distance);
+    }
+  }
+
+  // find the reason why a value was consumed at entry to a block
+  const ConsumedReason &findConsumedAtEntryReason(
+      SILBasicBlock *SILBlock, unsigned consumedVal) {
+    const BlockPartitionState &block = blockStates[SILBlock];
+    assert(block.getEntryPartition().isConsumed(consumedVal));
+
+    // check the cache
+    if (consumedAtEntryReasons.count({SILBlock, consumedVal}))
+      return consumedAtEntryReasons.at({SILBlock, consumedVal});
+
+    // enter a dummy value in the cache to prevent circular call dependencies
+    consumedAtEntryReasons[{SILBlock, consumedVal}] = ConsumedReason();
+
+    auto entryTracks = [&](unsigned val) {
+      return block.getEntryPartition().isTracked(val);
+    };
+
+    // this gets populated with all the tracked values at entry to this block
+    // that are consumed at the exit to some predecessor block, associated
+    // with the blocks that consume them
+    std::map<unsigned, std::vector<SILBasicBlock *>> consumedInSomePred;
+    for (SILBasicBlock *pred : SILBlock->getPredecessorBlocks())
+      for (unsigned consumedVal
+          : blockStates[pred].getExitPartition().getConsumedVals())
+        if (entryTracks(consumedVal))
+          consumedInSomePred[consumedVal].push_back(pred);
+
+    // this gets populated with all the multi-edges between values tracked
+    // at entry to this block that will be merged because of common regionality
+    // in the exit partition of some predecessor. It is not transitively closed
+    // because we want to count how many steps transitive merges require
+    std::map<unsigned, std::set<unsigned>> singleStepJoins;
+    for (SILBasicBlock *pred : SILBlock->getPredecessorBlocks())
+      for (std::vector<unsigned> region
+          : blockStates[pred].getExitPartition().getNonConsumedRegions()) {
+        for (unsigned fst : region) for (unsigned snd : region)
+            if (fst != snd && entryTracks(fst) && entryTracks(snd))
+              singleStepJoins[fst].insert(snd);
+      }
+
+    // this gets populated with the distance, in terms of single step joins,
+    // from the target consumedVal to other values that will get merged with it
+    // because of the join at entry to this basic block
+    std::map<unsigned, unsigned> distancesFromTarget;
+
+    //perform BFS
+    std::deque<std::pair<unsigned, unsigned>> processValues;
```

Comment:
no idea what these maps and deques are involving because distances and values are both `unsigned`. I think we need more than a typealias here.